gamemode/modules/base/sv_gamemode_functions.lua

```diff
@@ -312,34 +312,88 @@ for _, ply in pairs(player.GetHumans()) do
--- causing a lagfest when there are many players
-local function calcPlyCanHearPlayerVoice(listener)
-    if not IsValid(listener) then return end
-    DrpCanHear[listener] = DrpCanHear[listener] or {}
-    if listener:IsBot() then return end
-    local shootPos = listener:GetShootPos()
-    for _, talker in ipairs(player.GetHumans()) do
-        local talkerShootPos = talker:GetShootPos()
-        DrpCanHear[listener][talker] = not vrad or -- Voiceradius is off, everyone can hear everyone
-            (shootPos:DistToSqr(talkerShootPos) < voiceDistance and -- voiceradius is on and the two are within hearing distance
-                (not dynv or IsInRoom(shootPos, talkerShootPos, talker))) -- Dynamic voice is on and players are in the same room
-    end
-end
-hook.Add("PlayerInitialSpawn", "DarkRPCanHearVoice", function(ply)
-    calcPlyCanHearPlayerVoice(ply)
-    if ply:IsBot() then return end
-    timer.Create(ply:UserID() .. "DarkRPCanHearPlayersVoice", 0.5, 0, fn.Curry(calcPlyCanHearPlayerVoice, 2)(ply))
-end)
+local gridSize = GM.Config.voiceDistance*2 -- Grid cell size is equal to the size of the diamater of player talking
+local floor = math.floor -- Caching floor as we will need to use it a lot
+
+-- Grid based position check
+timer.Create("DarkRPCanHearPlayersVoice", 1, 0, function()
+    DrpCanHear = {}
+    local players = player.GetHumans()
+
+    -- Create Grid
+    local grid = {}
+    local plyToGrid = {
+        {},
+        {}
+    }
+
+    -- Get the grid position of every player O(N)
+    for _, ply in ipairs(players) do
+        local pos = ply:GetShootPos()
+        local x = floor(pos.x/gridSize)
+        local y = floor(pos.y/gridSize)
+        grid[x] = grid[x] or {}
+        grid[x][y] = grid[x][y] or {}
+        grid[x][y][#grid[x][y] + 1] = ply
+        plyToGrid[1][ply] = x
+        plyToGrid[2][ply] = y
+
+        DrpCanHear[ply] = {} -- Initialize output variable
+    end
+
+    -- Check all neighbouring cells for every player.
+    -- We are only checking in 1 direction to avoid duplicate check of cells
+    for _, ply1 in ipairs(players) do
+        local gridX = plyToGrid[1][ply1]
+        local gridY = plyToGrid[2][ply1]
+        local ply1ShootPos
+
+        for x = gridX, gridX+1 do   -- Get every neighbouring cell (including the actual cell) (x axis)
+            local row = grid[x]
+            if not row then continue end
+
+            for y = gridY, gridY+1 do   -- Get every neighbouring cell (including the actual cell) (y axis)
+                if x == gridX and y == gridY then continue end -- Exclude the actual cell as it needs a separate computation
+                local cell = row[y]
+                if not cell then continue end -- Check that values exist
+
+                -- Check if every player can talk
+                for _, ply2 in ipairs(cell) do
+                    ply1ShootPos = ply1ShootPos or ply1:GetShootPos()
```

Comment:
I don't think the ShootPos should be used for this since it's based on the model attachment - GetPos would be more suitable.

Reply:
Isn't GetPos at players' feet? Voices will be blocked by the tiniest of differences in height if that's the case.

Reply:
> Isn't GetPos at players' feet? Voices will be blocked by the tiniest of differences in height if that's the case.

It is, but it's comparing the distance between player's feet which is in a consistent position. GetShootPos can change to a large degree depending on the model and even more importantly, the rotation of the player. At least GetPos will never disqualify a player from hearing another based on them turning away from another.